### PR TITLE
Changes to support dynamic discretize / undo_discretize based on schema

### PR DESCRIPTION
**Pull request by hd23408**
The main difference is that it now reads the number of bins from the schema, and uses `numpy.linspace` to split into equal buckets within the min/max range. This is different from your original approach that was using arange, so you might want to double-check my logic there and make sure it does what you're expecting.

Additionally, I included a (manual) test case in the PR that just does a discretization / undo_discretization on the taxi data and outputs the results. This looks pretty similar to a discretize / undo_discretize using your original code, but you should obviously try this with your own test cases and see how it goes.

The schema used for discretize / undo_discretize is assumed to be a schema as created by https://github.com/hd23408/nist-schemagen (which is slightly different from the original taxi data schema). However, in most cases users will need to do a little manual tweaking of that schema after creating it. Specifically, these are the two things to look at, which you might want to include notes about in your own documentation:

1) In order to use the "clip" functionality, one of the columns in the schema MUST be marked as an "id" column ("kind" = "id") in the schema. This can only be done manually -- the schema generator does NOT attempt to identify an ID column. For your purposes, I think you would probably want to instruct the user to a) run the schema generator on the command line, b) edit the resultant parameters.json file to change the "kind" of the id column to "id", and then c) run your tool. (Note that this isn't necessary if you're not using "clip", though.)

2) Because the schema generator just spits out an arbitrary number of "bins" for each numeric column, the test case manually adjusts the number of bins for some of the columns to make them a little more sensical for the taxi data in particular. Again, this is something that a user would probably do manually in between generating the schema and running your synthetic data tool.

### CODE REVIEW

**Comment by ryan112358** [comment]
Looks good, thanks for the work on this.  A couple notes: 

1. I should probably do some reorganization of the repository, but currently the main directory contains the official submission made for the contest, and the extensions directory contains the new general-purpose mechanism we wrote.  With that said, I think it makes sense to keep util.py in the main directory unchanged, and include a new util.py in the extensions directory that can work with new datasets and slightly different input format expectations.  
2. Can you include in the PR an example schema exported by nist-schemagen as a json file so I can see what it looks like and so the code can run without that dependency if desired?
3. I think id columns and clipping will not be supported in the extension, so that is fine.

**Comment by hd23408** [comment]
Thanks, Ryan! I restored the main directory's util.py to the original, and moved the new util.py to the extensions directory. I also included a "parameters.json" file (in test/files/taxi_data/parameters.json) that is what the test case would produce for the taxi data specifically.

File: extensions/util.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        #    new[col] = np.minimum(new[col], info['max'])
+        #if 'min' in info: 
+        #    new[col] = np.maximum(new[col], info['min'])
+
+    dtypes = { col : schema[col]['dtype'] for col in schema }
+
+    return new.astype(dtypes)
+
+
+def score(real, synth):
```
Comment (ryan112358) [inline extensions/util.py:88]: I would delete this function, it's not relevant to other datasets (it hardcodes 'shift' and 'pickup_community_area' columns)

File: extensions/util.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+def downward_closure(cliques):
+    ans = set()
+    for proj in cliques:
+        ans.update(powerset(proj))
+    return list(sorted(ans, key=len))
+
+def discretize(df, schema, clip=None):
+
+    weights = None
+    if clip is not None:
```
Comment (ryan112358) [inline extensions/util.py:19]: let's remove clipping functionality I think, it didn't end up helping that much anyway

**Review by ryan112358 — CHANGES_REQUESTED** [review]

File: extensions/util.py (hunk 1 above)
Comment (hd23408) [reply extensions/util.py]: Done!

File: extensions/util.py (hunk 2 above)
Comment (hd23408) [reply extensions/util.py]: Okey dokey! "clip" has been removed.

**Pull request merged**